otherlibs/configurator/src/import.ml

```diff
@@ -0,0 +1,288 @@
+            acc
+          else
+            sub s ~pos:i ~len:(j - i) :: acc
+        in
+        List.rev acc
+      else
+        match s.[j] with
+        | '\r' -> loop ~last_is_cr:true ~acc i (j + 1)
+        | '\n' ->
+          let line =
+            let len =
+              if last_is_cr then
+                j - i - 1
+              else
+                j - i
+            in
+            sub s ~pos:i ~len
+          in
+          loop ~acc:(line :: acc) (j + 1) (j + 1) ~last_is_cr:false
+        | _ -> loop ~acc i (j + 1) ~last_is_cr:false
+    in
+    loop ~acc:[] 0 0 ~last_is_cr:false
+
+  let exists =
+    let rec loop s i len f =
+      if i = len then
+        false
+      else
+        f (unsafe_get s i) || loop s (i + 1) len f
+    in
+    fun s ~f -> loop s 0 (length s) f
+
+  let is_empty = function
+    | "" -> true
+    | _ -> false
+
+  let extract_words s ~is_word_char =
+    let rec skip_blanks i =
+      if i = length s then
+        []
+      else if is_word_char s.[i] then
+        parse_word i (i + 1)
+      else
+        skip_blanks (i + 1)
+    and parse_word i j =
+      if j = length s then
+        [ sub s ~pos:i ~len:(j - i) ]
+      else if is_word_char s.[j] then
+        parse_word i (j + 1)
+      else
+        sub s ~pos:i ~len:(j - i) :: skip_blanks (j + 1)
+    in
+    skip_blanks 0
+
+  let extract_comma_space_separated_words s =
+    extract_words s ~is_word_char:(function
+      | ','
+      | ' '
+      | '\t'
+      | '\n' ->
+        false
+      | _ -> true)
+
+  let extract_blank_separated_words s =
+    extract_words s ~is_word_char:(function
+      | ' '
+      | '\t' ->
+        false
+      | _ -> true)
+
+  let split s ~on =
+    let rec loop i j =
+      if j = length s then
+        [ sub s ~pos:i ~len:(j - i) ]
+      else if s.[j] = on then
+        sub s ~pos:i ~len:(j - i) :: loop (j + 1) (j + 1)
+      else
+        loop i (j + 1)
+    in
+    loop 0 0
+end
+
+module Io = struct
+  let open_in ?(binary = true) fn =
+    if binary then
+      Stdlib.open_in_bin fn
+    else
+      Stdlib.open_in fn
+
+  let open_out ?(binary = true) fn =
+    if binary then
+      Stdlib.open_out_bin fn
+    else
+      Stdlib.open_out fn
+
+  let input_lines =
+    let rec loop ic acc =
+      match input_line ic with
+      | exception End_of_file -> List.rev acc
+      | line -> loop ic (line :: acc)
+    in
+    fun ic -> loop ic []
+
+  let with_file_in ?binary fn ~f =
+    Exn.protectx (open_in ?binary fn) ~finally:close_in ~f
+
+  let with_file_out ?binary p ~f =
+    Exn.protectx (open_out ?binary p) ~finally:close_out ~f
+
+  let write_file ?binary fn data =
+    with_file_out ?binary fn ~f:(fun oc -> output_string oc data)
+
+  let write_lines ?binary fn lines =
+    with_file_out ?binary fn ~f:(fun oc ->
+        List.iter
+          ~f:(fun line ->
+            output_string oc line;
+            output_string oc "\n")
+          lines)
+
+  let read_all =
```

Comment:
Note that we could use the simple version here. We only use `read_file` to read temporary files we created ourselves in configurator, so the simple version will always be enough.

Reply:
That's true but I find that just copying the implementation directly from stdune makes maintenance easier. We can just copy/paste to update the code without worrying about little details

Reply:
That's true. BTW, the Io module (at least the String version) feels like something that could easily be cleaned up and stabilised. So we could mint some library for it to share the code.